diff -Naur nginx-1.16.1/src/http/ngx_http.c nginx-1.16.1_patch/src/http/ngx_http.c
--- nginx-1.16.1/src/http/ngx_http.c 2019-08-13 20:51:43.000000000 +0800
+++ nginx-1.16.1_patch/src/http/ngx_http.c 2019-11-04 18:08:45.038900915 +0800
@@ -71,6 +71,7 @@

ngx_http_output_header_filter_pt ngx_http_top_header_filter;
ngx_http_output_body_filter_pt ngx_http_top_body_filter;
+ngx_http_output_write_filter_pt ngx_http_top_write_filter;
ngx_http_request_body_filter_pt ngx_http_top_request_body_filter;

diff -Naur nginx-1.16.1/src/http/ngx_http_core_module.h nginx-1.16.1_patch/src/http/ngx_http_core_module.h
--- nginx-1.16.1/src/http/ngx_http_core_module.h 2019-08-13 20:51:43.000000000 +0800
+++ nginx-1.16.1_patch/src/http/ngx_http_core_module.h 2019-11-04 18:08:47.526961215 +0800
@@ -511,6 +511,8 @@
typedef ngx_int_t (*ngx_http_output_header_filter_pt)(ngx_http_request_t *r);
typedef ngx_int_t (*ngx_http_output_body_filter_pt)
(ngx_http_request_t *r, ngx_chain_t *chain);
+typedef ngx_int_t (*ngx_http_output_write_filter_pt)

    (ngx_http_request_t *r, off_t bsent);
    typedef ngx_int_t (*ngx_http_request_body_filter_pt)
    (ngx_http_request_t *r, ngx_chain_t *chain);

diff -Naur nginx-1.16.1/src/http/ngx_http.h nginx-1.16.1_patch/src/http/ngx_http.h
--- nginx-1.16.1/src/http/ngx_http.h 2019-08-13 20:51:43.000000000 +0800
+++ nginx-1.16.1_patch/src/http/ngx_http.h 2019-11-04 18:08:47.156952249 +0800
@@ -174,6 +174,7 @@

extern ngx_http_output_header_filter_pt ngx_http_top_header_filter;
extern ngx_http_output_body_filter_pt ngx_http_top_body_filter;
+extern ngx_http_output_write_filter_pt ngx_http_top_write_filter;
extern ngx_http_request_body_filter_pt ngx_http_top_request_body_filter;

diff -Naur nginx-1.16.1/src/http/ngx_http_write_filter_module.c nginx-1.16.1_patch/src/http/ngx_http_write_filter_module.c
--- nginx-1.16.1/src/http/ngx_http_write_filter_module.c 2019-08-13 20:51:43.000000000 +0800
+++ nginx-1.16.1_patch/src/http/ngx_http_write_filter_module.c 2019-11-04 18:08:48.129975830 +0800
@@ -293,6 +293,9 @@
return NGX_ERROR;
}

    if (ngx_http_top_write_filter){

       ngx_http_top_write_filter(r, sent);

    }
    if (r->limit_rate) {

    nsent = c->sent;


